### PR TITLE
MAT-1253: Bor tests

### PR DESCRIPTION
**Pull request by tokamak-git**

### CODE REVIEW

File: bor/side_handler_test.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	for i, c := range tc {
+		suite.SetupTest()
+		c.msg = fmt.Sprintf("i: %v, msg: %v", i, c.msg)
+		if c.cm != nil {
+			for _, m := range c.cm {
+				suite.mockCaller.On(m.name, m.args...).Return(m.ret)
+			}
+		}
+
+		out := bor.SideHandleMsgSpan(suite.ctx, suite.app.BorKeeper, borTypes.MsgProposeSpan{}, &suite.mockCaller)
```
Comment (ptsayli) [inline bor/side_handler_test.go:76]:
```suggestion
out:= suite.sideHandler(suite.ctx, msgProposeSpan)
require.Equal(t, abci.SideTxResultType_Skip, out.Result, "Result should bae `skip`")
```

File: bor/side_handler_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		// construct output
+		c.out = abci.ResponseDeliverSideTx{Code: uint32(c.code), Codespace: c.codespace, Result: abci.SideTxResultType_Skip}
```
Comment (ptsayli) [inline bor/side_handler_test.go:78]: can remove this

File: bor/handler_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		if c.span != nil {
+			suite.app.BorKeeper.AddNewSpan(suite.ctx, *c.span)
+		}
+
+		// cSpan is used to check if span data remains constant post handler execution
+		cSpan := suite.app.BorKeeper.GetAllSpans(suite.ctx)
+
+		out := bor.HandleMsgProposeSpan(suite.ctx, borTypes.MsgProposeSpan{ID: c.msgID, Proposer: c.proposer, StartBlock: c.startBlock, EndBlock: c.endBlock, ChainID: c.chainID, Seed: c.seed}, suite.app.BorKeeper)
+		suite.Equal(c.out, out, c.msg)
```
Comment (ptsayli) [inline bor/handler_test.go:137]: can change to `require.True(t, !out.IsOK(), errs.CodeToDefaultMsg(out.Code))`

File: bor/handler_test.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+			chainID:    "15001", // default chain id
+			startBlock: 1,
+			endBlock:   1,
+			msgID:      2,
+			out:        common.ErrInvalidSpanDuration("1").Result(),
+			msg:        "error span invalid duration",
+		},
+	}
+	for i, c := range tc {
+		suite.SetupTest()
```
[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/handler_test.go:124]: Am running multiple tests within the loop for different values. This way I do not have to rewrite all the setup functions for every test case. the `suite.SetupTest` allows for me to refresh the values before every test case

File: bor/keeper_test.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+			id:         1,
+			startBlock: 3,
+			msg:        "validation error missing id",
+		},
+	}
+
+	for i, c := range tc {
+		cMsg := fmt.Sprintf("i: %v, msg: %v", i, c.msg)
+		err := suite.app.BorKeeper.FreezeSet(suite.ctx, c.id, c.startBlock, c.endBlock, c.borChainID, c.seed)
+		suite.Equal(c.expErr, err, cMsg)
```
Comment (mankenavenkatesh) [inline bor/keeper_test.go:58]: Can we also check if span is stored or not, 
lastEthBlock is incremented or not.

File: bor/keeper_test.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+		},
+	}
+
+	for i, c := range tc {
+		suite.app.BorKeeper.SetParams(suite.ctx, bortypes.Params{SprintDuration: 1, SpanDuration: 1, ProducerCount: c.producerCount})
+		simulation.LoadValidatorSet(4, suite.T(), suite.app.StakingKeeper, suite.ctx, true, 0) // load a random number of validators
+		cMsg := fmt.Sprintf("i: %v, msg: %v", i, c.msg)
+		out, err := suite.app.BorKeeper.SelectNextProducers(suite.ctx, c.seed)
+		suite.NotNil(out, cMsg)
+		suite.Equal(c.expErr, err, cMsg)
```
Comment (mankenavenkatesh) [inline bor/keeper_test.go:130]: only nil and err check ??  
we need to check if expected producers count coming

File: app/app.go [hunk 7]
```diff
@@ -498,6 +498,10 @@ func MakeCodec() *codec.Codec {
 	return cdc
 }
 
+func (app *HeimdallApp) GetCaller() helper.ContractCaller {
```
[the remark this answers is not in the document]
Comment (tokamak-git) [reply app/app.go:502]: do not have access to the unexported fields in app object, thus require caller

File: bor/side_handler_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+func TestBorSideChHandlerTestSuite(t *testing.T) {
+	suite.Run(t, new(sideChHandlerSuite))
+}
+
+func (suite *sideChHandlerSuite) SetupTest() {
+	isCheckTx := false
+	suite.app = app.Setup(isCheckTx)
+	suite.ctx = suite.app.BaseApp.NewContext(isCheckTx, abci.Header{})
+	suite.mockCaller = mocks.IContractCaller{}
```
[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/side_handler_test.go:41]: directly testing the functions within as they allow for more control on the values Im passing...

File: bor/side_handler_test.go (hunk 2 above)
Comment (tokamak-git) [reply bor/side_handler_test.go]: using it to set the values for the output, shall refine it...

File: bor/handler_test.go (hunk 3 above)
Comment (tokamak-git) [reply bor/handler_test.go:138]: want to be able to check the exact response, this way I am able to verify both errors and happy flow within the same test func...

File: bor/handler_test.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+			suite.app.BorKeeper.SetParams(suite.ctx, borTypes.Params{SpanDuration: c.spanDuration})
+		}
+		if c.span != nil {
+			suite.app.BorKeeper.AddNewSpan(suite.ctx, *c.span)
+		}
+
+		// cSpan is used to check if span data remains constant post handler execution
+		cSpan := suite.app.BorKeeper.GetAllSpans(suite.ctx)
+
+		out := bor.HandleMsgProposeSpan(suite.ctx, borTypes.MsgProposeSpan{ID: c.msgID, Proposer: c.proposer, StartBlock: c.startBlock, EndBlock: c.endBlock, ChainID: c.chainID, Seed: c.seed}, suite.app.BorKeeper)
```
[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/handler_test.go:137]: makes it easier to test for a bunch of errors and mock the calls, shall be testing the handler separately...

File: bor/handler_test.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func TestBorHandlerTestSuite(t *testing.T) {
+	suite.Run(t, new(handlerSuite))
+}
+
+func (suite *handlerSuite) SetupTest() {
+	isCheckTx := false
+	suite.app = app.Setup(isCheckTx)
+	suite.ctx = suite.app.BaseApp.NewContext(isCheckTx, abci.Header{})
```
[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/handler_test.go:32]: makes it easier to test for a bunch of errors and mock the calls, shall be testing the handler separately...

File: bor/keeper_test.go (hunk 6 above)
Comment (tokamak-git) [reply bor/keeper_test.go:130]: not sure if that can be achieved, have set a seed yet it generates a new list everytime...

File: bor/handler_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ptsayli) [reply bor/handler_test.go:32]: cool, then may be we can add testcase for `NewHandler()`

File: app/app.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ptsayli) [reply app/app.go:502]: Please remove `GetCaller()`, we are not using it anywhere

File: bor/keeper_test.go [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+			seed:          common.HexToHash("testSeed"),
+			msg:           "happy flow with fewer validators than allowed",
+			out:           simulation.LoadValidatorSet(4, suite.T(), suite.app.StakingKeeper, suite.ctx, true, 0), // load a random number of validators
+		},
+	}
+
+	for i, c := range tc {
+		suite.SetupTest()
+		// cVals is used to check if validators are being modified during execution
+		cVals := suite.app.StakingKeeper.GetValidatorSet(suite.ctx)
```
[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/keeper_test.go:155]: validators are being loaded, I am validating the objects are not empty. lets discuss this offline...

**Pull request merged**